library/core/src/fmt/mod.rs

```diff
@@ -231,6 +231,133 @@ impl<W: Write + ?Sized> Write for &mut W {
     }
 }
 
+/// A `WriteCursor` implements [`fmt::Write`] by writing to an in-memory buffer.
+///
+/// [`fmt::Write`]: self::Write
+///
+/// # Examples
+///
+/// ```
+/// #![feature(fmt_write_cursor)]
+/// use std::fmt::WriteCursor;
+///
+/// # fn test_write_cursor() -> std::fmt::Result {
+/// let mut buf = [0u8; 5];
+/// let mut cursor = WriteCursor::new(&mut buf);
+/// write!(cursor, "{}", 12345)?;
+/// assert_eq!(cursor.as_str(), "12345");
+/// assert_eq!(&buf, b"12345");
+/// # Ok(())
+/// # }
+/// ```
+#[doc(alias = "sprintf")]
+#[doc(alias = "snprintf")]
+#[unstable(feature = "fmt_write_cursor", issue = "none")]
+pub struct WriteCursor<'a> {
+    buf: &'a mut [mem::MaybeUninit<u8>],
+    utf8_len: usize,
+}
+
+#[unstable(feature = "fmt_write_cursor", issue = "none")]
+impl Debug for WriteCursor<'_> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        f.debug_struct("WriteCursor")
+            .field("capacity", &self.capacity())
+            .field("written", &self.written())
+            .finish()
+    }
+}
+
+impl<'a> WriteCursor<'a> {
+    /// Creates a new cursor wrapping the provided buffer.
+    #[unstable(feature = "fmt_write_cursor", issue = "none")]
+    pub fn new(buf: &'a mut [u8]) -> WriteCursor<'a> {
+        // SAFETY: [T] and [MaybeUninit<T>] have the same layout
+        Self::new_uninit(unsafe { mem::transmute(buf) })
+    }
+
+    /// Creates a new cursor wrapping the provided uninitialized buffer.
+    #[unstable(feature = "fmt_write_cursor", issue = "none")]
+    pub fn new_uninit(buf: &'a mut [mem::MaybeUninit<u8>]) -> WriteCursor<'a> {
+        WriteCursor { buf, utf8_len: 0 }
+    }
+
+    /// Returns the total capacity of the cursor's underlying buffer, in bytes.
+    #[unstable(feature = "fmt_write_cursor", issue = "none")]
+    pub fn capacity(&self) -> usize {
+        self.buf.len()
+    }
+
+    /// Returns how many bytes have been written to the cursor's underlying buffer.
+    #[unstable(feature = "fmt_write_cursor", issue = "none")]
+    pub fn written(&self) -> usize {
+        self.utf8_len
+    }
+
+    /// Returns the written portion of the cursor's underlying buffer as a
+    /// byte slice.
+    ///
+    /// The returned slice contains valid UTF-8.
+    #[unstable(feature = "fmt_write_cursor", issue = "none")]
+    pub fn as_bytes(&self) -> &[u8] {
+        let utf8 = &self.buf[..self.utf8_len];
+        // SAFETY: The buffer is incrementally initialized by `write_str`, which
+        // updates `utf8_len`.
+        unsafe { mem::MaybeUninit::slice_assume_init_ref(utf8) }
+    }
+
+    /// Consumes the cursor and returns the written portion of the cursor's
+    /// underlying buffer as a byte slice.
+    ///
+    /// The returned slice contains valid UTF-8.
+    #[unstable(feature = "fmt_write_cursor", issue = "none")]
+    pub fn into_bytes(self) -> &'a [u8] {
```

Comment:
maybe the return type should be `&'a mut [u8]`?

Reply:
Great idea! Done.